Feature/pro 647 pillar staking contract

cryptonoyaiba opened the pull request:
- Staking contract
- Test suite for staking contract
- Deployment script for staking contract
- Documentation for staking contract

cryptonoyaiba committed:
WIP staking contract and testing

- Started work on the staking contract and test suite

cryptonoyaiba committed:
Added modifiers to min and max state updates

- Only allows min and max state updates when contract state is STAKEABLE
- Updated and added new tests for test suite for these changes

cryptonoyaiba committed:
Added staking and reward token changes

- Added staking and reward tokens
- Added mapping for rewardBalances
- Updated error for contract state to be clearer
- Updated some function visibility modifiers
- Updated test suite and added new tests

cryptonoyaiba committed:
Staking contract

- Staking contract completed
- Added test suite for staking contract
- Added deployment script for staking contract
- Added documentation for staking contract
- Audited through manual audit and Slither
- Removed errors from separate contract

cryptonoyaiba committed:
Merge branch 'main' into feature/PRO-647-pillar-staking-contract

ch4r10t33r commented on contracts/testing_utils/DummyPillarToken.sol:
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
```
You wouldn't need this to be part of the repo.

lbw33 replied:
This is required for testing.

cryptonoyaiba replied:
Required for testing

ch4r10t33r commented on contracts/testing_utils/DummyWETHToken.sol:
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
```
This isn't required as well

cryptonoyaiba replied:
This is also require for testing.

ch4r10t33r replied:
There are several contract libraries that you can use for deploying a dummy token, no? Anyway as long its just for testing, this is fine.

ch4r10t33r commented on package.json:
```diff
@@ -1,18 +1,22 @@
-    "dotenv": "14.2.0",
-    "hardhat-gas-reporter": "1.0.7"
-  }
+    "name": "pillar-dao-contracts",
```
Please use exact version of the packages.

cryptonoyaiba replied:
Done

ch4r10t33r commented on scripts/01_deploy_pillar_staking.js:
```diff
@@ -0,0 +1,39 @@
+const main = async () => {
+  await run("compile");
+
+  const stakingToken = "0xe3818504c1B32bF1557b16C238B2E01Fd3149C17"; // add staking token address here (PLR)
```
These contract addresses are valid only on Ethereum Mainnet. So please add a necessary check to validate the chainId and then use these addresses.

cryptonoyaiba replied:
Done

ch4r10t33r commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,271 @@
+        if (stakingToken.balanceOf(msg.sender) < _amount)
+            revert InsufficientBalance();
+        totalStaked = totalStaked + _amount;
+        stakeholderData[msg.sender].stakedAmount =
```
Please handle a scenario where a user can stake any amount between minStake and maxStaje multiple times, his cumulative totalStake shouldn't exceed the maxStake.

cryptonoyaiba replied:
That is covered in the revert above this:
```
if (_amount > maxStake - stakeholderData[msg.sender].stakedAmount)
            revert StakeWouldBeGreaterThanMax();
```

ch4r10t33r replied:
Thanks. Pls add a specific tests for this.

ch4r10t33r commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,271 @@
+        whenRewardsAllocated
+    {
+        uint256 stakedBalance = stakeholderData[msg.sender].stakedAmount;
+        totalStaked = totalStaked - stakedBalance;
```
Please use SafeMath to ensure there are no overflows.

cryptonoyaiba replied:
SafeMath isnt needed in sol v0.8.0+ its in built

ch4r10t33r replied:
Ah, I didn't notice your compiler version.

ch4r10t33r commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,271 @@
+
+    /* ========== RESTRICTED FUNCTIONS ========== */
+
+    function calculateRewardAllocation() external onlyOwner {
```
We need a specific contract function to allocate reward tokens. This can be involved only by the owner. Allocation of rewards should trigger an Event.

cryptonoyaiba replied:
I am not sure what you mean. This fn allocates rewards to users and is invoked by only the owner. This calculates the reward amount and stores it against the account. Then when the user withdraws their stake it gives them their rewards. There is an event for reward allocation and an event for rewards granted (inside the unstake fn)

ch4r10t33r replied:
The function is used to transfer WETH to the contract (however instead of just transferring the rewards, we call a specific contract function notifying that the rewards have been added to the staking contract and we are ready for reward distribution).

cryptonoyaiba replied:
Added in a fn to deposit rewards tokens that emits and event. Also added a check in to make sure it is the rewards that have been deposited by the contract owner and not from another source.

ch4r10t33r commented on test/PillarStaking.spec.js:
```diff
@@ -0,0 +1,617 @@
+    });
+  });
+
+  describe("Staking PLR", () => {
```
Add a specific condition to test the same user staking multiple times during the staking period.

cryptonoyaiba replied:
Done

cryptonoyaiba committed:
Locked packages, added tests and amended deployment

- Locked package versions
- Added depositRewards fn to allow owner to deposit reward tokens to the contract.
- Added check to see if reward tokens are deposited as this could potentially be manipulated by a malicious actor transferring the reward token into the contract without using the depositRewards fn that would satisfy the conditions used on rewarding tokens.
- Added test case for same user staking multiple times in the staking period and for depositRewards fn and amended test cases to use this new fn
- Added logic to deployment to detect chain id
- Updated docs

cryptonoyaiba committed:
Merge branch 'feature/PRO-647-pillar-staking-contract' of https://github.com/pillarwallet/pillar-dao-contracts into feature/PRO-647-pillar-staking-contract

ch4r10t33r commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,281 @@
+        if (_amount == 0) revert RewardsCannotBeZero();
+        rewardToken.safeTransferFrom(msg.sender, address(this), _amount);
+        rewardsDeposited = true;
+        emit RewardsDeposited();
```
Shouldn't the event report the amount of rewards that has been deposited?

cryptonoyaiba replied:
Done

cryptonoyaiba committed:
Added rewards to rewardsDeposited event

- Added rewards param to rewardsDeposited event
- Amended test to check for new param
- Removed extra compiler block from hardhat config

0xSulpiride commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,281 @@
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+
+import "hardhat/console.sol";
```
>import "hardhat/console.sol"

Do we need this?

cryptonoyaiba replied:
ooops i'll remove that

0xSulpiride commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,281 @@
+    function calculateRewardAllocation() external onlyOwner {
+        if (!rewardsDeposited) revert RewardsNotTransferred();
+        uint256 totalRewards = rewardToken.balanceOf(address(this));
+        for (uint256 i; i < stakeholderList.length; ++i) {
```
Pls add a function that rewards a specific user and a function to get the list of users.

This is in case there will be too many users in the `stakeholdersList` and we might not be able to reward all of them because it would hit the block gas limit.

0xSulpiride replied:
Or maybe we should add the `claim` function, so users will pay for the calculation of their rewards themselves?

0xSulpiride replied:
If users should not be able to claim rewards for themselves, I think you could add parameters for the`calculateRewardAllocation` function to set which partition of the `stakeholderList` to reward, so we could divide the process into multiple transactions.

cryptonoyaiba replied:
I suppose I could separate this to work for a specific address and have that built into the unstake fn. I will need to rethink how the logic will work for this

cryptonoyaiba replied:
What I am thinking is rather than calculating rewards for all users in one function, have a function that calculates the rewards for 1 user and build this into the unstake function so that the user will pay for the calculation of their own rewards. This amount will be stored against the user so that they can see how much they should be paid. The only issue with this is that before unstaking (and paying out of rewards), the user will not know how much rewards are due to them

0xSulpiride replied:
>have a function that calculates the rewards for 1 user and build this into the unstake function so that the user will pay for the calculation of their own rewards

Yeah, I think this would be better.

>The only issue with this is that before unstaking (and paying out of rewards), the user will not know how much rewards are due to them

I'm not quite getting this part 🤔 Rewards are fixed after the lock period begins, no?

cryptonoyaiba replied:
No as we dont know what the amount they are due would be as the rewards are unknown. Their % of rewards would be know but not the amount.

ch4r10t33r replied:
This is a good point @marie-fourier . I would say the reward calculation logic should be part of the unstake function. We determine how much WETH a user is eligible for at the time of unstake and let that be transferred to him.

0xSulpiride commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,281 @@
+        external
+        nonReentrant
+        whenReadyForUnstake
+        whenRewardsAllocated
```
Don't we allow unstaking before the lock period? Just curious

cryptonoyaiba replied:
No. Staking is locked for 12+ months and until the rewards have been deposited and calculated

0xSulpiride reviewed (changes requested):
Great job! I have some minor comments

ch4r10t33r commented:
Btw, I am just wondering if we should issue a new stPLR token equivalent to the number of a user has staked into the contract? and we burn these stPLR tokens while claiming?

cryptonoyaiba committed:
Added changes to how rewards are calculated

- Rewards now calculated on a per user basis upon calling the unstake fn. This allows the fee associated with calculating rewards to be passed onto the user who is staking.
- Amended test suite for new code and added new tests for added functionality
- Amended docs to reflect changes

0xSulpiride commented on contracts/PillarStaking.sol:
```diff
@@ -162,35 +155,24 @@ contract PillarStaking is ReentrancyGuard, Ownable {
         if (_amount == 0) revert RewardsCannotBeZero();
         rewardToken.safeTransferFrom(msg.sender, address(this), _amount);
-        rewardsDeposited = true;
+        rewards = _amount;
```
pls make this `rewards += _amount`

cryptonoyaiba replied:
Done

0xSulpiride replied:
thanks. The name `deposit` implies that the function might be called multiple times

cryptonoyaiba replied:
Do you have any suggestions for a rename? I can't think of what else to call it as it is depositing rewards to the contract.

0xSulpiride replied:
I think `setRewards` would a better naming

0xSulpiride commented on contracts/PillarStaking.sol:
```diff
@@ -162,35 +155,24 @@ contract PillarStaking is ReentrancyGuard, Ownable {
-        }
-        rewardsAllocated = true;
-        emit RewardsAllocated();
+    function calculateRewardAllocation(address _staker)
```
can we make this function `external whenReadyForUnstake` so users will be able to see their rewards?

cryptonoyaiba replied:
And remove it from inside of the unstake fn?

cryptonoyaiba replied:
I would have to add some logic to check to see whether the user has already called this in the unstake fn (it would make no sense in them having to incur the cost of calling this twice) and if they call it outside of the unstake fn then we will have to store it against their address. I think we should do it this way then to make sure that if the user doesnt check their rewards, it will still calculate them inside the unstake fn.

0xSulpiride replied:
ah, right, I meant `public whenReadyForUnstake`

0xSulpiride replied:
actually, if it's not the requirement, then let's skip it @lbw33

cryptonoyaiba replied:
No I think that it is a good idea. It would be good for users to see their reward allocation before they unstake.

cryptonoyaiba replied:
Done

cryptonoyaiba committed:
Created PillarStaking token and made amendments to staking contract

- Created PillarStaking token: Gets minted to users upon stake and burned upon unstake. Represents user's stake on 1:1 basis.
- Created interface for PillarStaking token.
- Amended staking contract to allow users to check their reward allocation before unstaking. Changed depositRewards fn to setRewards and changed/added some errors/events.
- Amended test suite to reflect for changed make to staking contract and added tests to check that PillarStaking tokens are minted/burned.
- Amended deployment script to reflect new deployment parameters.
- Updated documentation to reflect changes and added some documentation for PillarStaking token.

0xSulpiride commented on contracts/PillarStaking.sol:
```diff
@@ -155,24 +182,21 @@ contract PillarStaking is ReentrancyGuard, Ownable {
-        uint256 stakedPercentage = ((stakeAmount * BPS) / totalStaked);
-        uint256 reward = ((stakedPercentage * rewards) / BPS);
-        return reward;
+        rewards += _amount;
```
sorry, my previous comment was ambiguous. I meant that the `depositRewards` implies that here should be `rewards += _amount`, but if we're renaming this function to `setRewards` then it's better to write `rewards = _amount` imo

cryptonoyaiba replied:
Ok i'll change the fn name back to depositRewards as its more descriptive of what the fn does.

cryptonoyaiba replied:
Done

0xSulpiride reviewed (approved):
LGTM. just one comment

cryptonoyaiba committed:
Changed setRewards fn to depositRewards

0xSulpiride reviewed (approved):
LGTM

ch4r10t33r commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,289 @@
+        if (_maxTotalStake > 0) maxTotalStake = _maxTotalStake;
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+        stakedToken = IPillarStakingToken(_stakedToken);
```
Shouldn't we be instantiating a new StakingToken here rather than using an already deployed token?

Given that you are using AccessControl, if we are to deploy the contract from the PillarStaking contract then only the contract can mint/burn the tokens, no?

cryptonoyaiba replied:
No problem I can integrate that into the staking contract

cryptonoyaiba replied:
Done.

ch4r10t33r commented on contracts/PillarStaking.sol:
```diff
@@ -0,0 +1,289 @@
+        public
+        whenReadyForUnstake
+    {
+        if (stakeholderData[msg.sender].rewardAmount != 0)
```
Please add a check for a zero address here. Also make this internal function.

cryptonoyaiba replied:
So this was how it was originally however, Utkir raised the point that users will never be able to check their reward allocation before the rewards are transferred to them. So this should be reverted back to it original state of being baked into the unstake fn and that is it?

ch4r10t33r replied:
@lbw33 I am talking about checking if the passed address parameter isn't zero. If a user wishes to check his reward balance then we shouldn't be publishing an event within this function! You can move the event to the unstake function.

cryptonoyaiba replied:
Ok, i'll put in a zero address check and remove the event. Are you happy for this to still remain a public function?

cryptonoyaiba committed:
Staking contract changes

- PillarStakedToken now gets deployed from staking contract constructor and automatically set as MINTER_ROLE.
- Changed allocation of rewards to no longer emit and event and added zero address check.
- Updated test suite to reflect changes.
- Updated docs to reflect changes.

cryptonoyaiba committed:
Added zero address test for calculateRewardAllocation

cryptonoyaiba committed:
Amendments to contracts

- Changed `calculateRewardAllocation` to `eligibleRewardAmount`.
- Changed the `DEFAULT_ADMIN_ROLE` from `msg.sender` to `tx.origin` to allow for re use of stake token.
- Updated test suite and docs to reflect changes.

cryptonoyaiba committed:
Added transfer lock to stkPLR token

- Override ERC20 transfer function to add modifier to only let staking token transfer tokens. Solves issue of transferring stkPLR tokens causing issues with unstake fn.

cryptonoyaiba committed:
Added fix for balance adjustments after unstake and amended deployment script

- Added fix for balances not zeroing after unstaking.
- Added ability to deploy tokens for testing in deployment script.
